pkg/storage/kubernetes/ipam.go

```diff
@@ -378,20 +378,20 @@ func IPManagement(mode int, ipamConf whereaboutstypes.IPAMConfig, containerID st
 
 	go func() {
 		defer wg.Done()
-		ctx, cancel := context.WithCancel(context.Background())
 		res := make(chan error)
+		leCtx, leCancel := context.WithCancel(context.Background())
```

Comment:
should it be the parent ctx instead of context.Background? 
How leCtx will honor the timeout set by top level ctx?

Reply:
Good catch @rthakur-est ! IMO, this should be `leCtx, leCancel := context.WithCancel(ctx)` (the parent ctx).

Reply:
IMO, the top level context timeout is not required. Let the leader election run with indefinite timeout as is done currently and use releaseOnCancel from #145 carefully with request timeout ctx. If leader lease cancellation is done only after ippool update, that will ensure that we do not end up with duplicate IPs and faster relinquishing of lease will reduce the overall time.

Reply:
> You've renamed this variable to stop shadowing the context from the function's input parameter, right ?

Correct and to make it slightly easier to read the context is associated with.

> should it be the parent ctx instead of context.Background?

I thought long about this. I do not think it should be parent process. If the parent process killed the LE thread, then we cannot guarantee that we weren't currently acting on the backend in ```IPManagementKubernetesUpdate```. I want to exit out of that function first and then we can end the ```le.Run's``` context.

> How leCtx will honor the timeout set by top level ctx?

It is indirectly honoured. When top level ctx ends, first we return from:
```
	go func() {
		defer wg.Done()
		for {
			select {
			case <-ctx.Done():
				err = fmt.Errorf("time limit exceeded while waiting to become leader")
				stopM <- struct{}{}
				return
			case <-leader:
				logging.Debugf("Elected as leader, do processing")
				newip, err = IPManagementKubernetesUpdate(ctx, mode, ipam, ipamConf, containerID, podRef)
				stopM <- struct{}{}
				return
			case <-deposed:
				logging.Debugf("Deposed as leader, shutting down")
				result <- nil
				return
			}
		}
	}()
```

This allows us to be sure we have stopped working on the backend in ```IPManagementKubernetesUpdate```.
Returning from the go routine above causing the cancellation of ```le.Run()```s context. So, we honor the top level context indirectly.
I know its horribly confusing.. and that is how bugs are hidden.

> Good catch @rthakur-est ! IMO, this should be leCtx, leCancel := context.WithCancel(ctx) (the parent ctx).

I disagree. If we do this and if ```ReleaseOnCancel``` is enabled, we give give rise to a race condition in ```IPManagementKubernetesUpdate``` and some other whereabouts instance that can instantly get the lease lock. I want to make sure we are out of ```IPManagementKubernetesUpdate``` before stoppling LE (and releasing the lock).

> IMO, the top level context timeout is not required. Let the leader election run with indefinite timeout as is done currently and use releaseOnCancel from #145 carefully with request timeout ctx. If leader lease cancellation is done only after ippool update, that will ensure that we do not end up with duplicate IPs and faster relinquishing of lease will reduce the overall time.

The reason I have a top level context is to prevent multiple instances of wb when kubelet respawns another instance if the pods status field isn't populated. I also kept this outside the function so I could also have a timeout on cmdDel.

Reply:
> Kubelet will respawn another instance only if it receives an error from cni plugin

Not only, if the CNI plugin doesn't respond within a set time, kubelet will check the "status" field, and try to create the pod again.

Reply:
I wasn't aware of this. In that case, this timeout can be kept close to the timeout that kubelet has.

Reply:
youll see it happen at scale. If you want to repro it locally, edit the wb code and put a long sleep in the cmdAdd path. Look at kubelet logs and when it recreates the pod, check the number of wb instances.